### ISSUE TITLE
k8s api token not auto refreshing (aws eks warning)

### ISSUE THREAD

**Issue by ldemailly** [kiali/kiali#5070]
Here is what we got from amazon with arn:aws:eks:XXXXX|istio-system:kiali (xxx is our account/region/clustername details) being the kiali 1.48 pod - a colleague upgraded to 1.50 and still see the warnings

What is surprising is that the k8s api client I see in go.mod seems above the version they mention as min required... 

----
Hello,

We have identified applications running in one or more of your Amazon EKS clusters that are not refreshing service account tokens. Applications making requests to Kubernetes API server with expired tokens will fail. You can resolve the issue by updating your application and its dependencies to use newer versions of Kubernetes client SDK that automatically refreshes the tokens.

What is the problem?

Kubernetes version 1.21 graduated BoundServiceAccountTokenVolume feature [1] to beta and enabled it by default. This feature improves security of service account tokens by requiring a one hour expiry time, over the previous default of no expiration. This means that applications that do not refetch service account tokens periodically will receive an HTTP 401 unauthorized error response on requests to Kubernetes API server with expired tokens. You can learn more about the BoundServiceAccountToken feature in EKS Kubernetes 1.21 release notes [2].

To enable a smooth migration of applications to the newer time-bound service account tokens, EKS v1.21+ extends the lifetime of service account tokens to 90 days. Applications on EKS v1.21+ clusters that make API server requests with tokens that are older than 90 days will receive an HTTP 401 unauthorized  error response.

How can you resolve the issue?

To make the transition to time bound service account tokens easier, Kubernetes has updated the below official versions of client SDKs to automatically refetch tokens before the one hour expiration:

* Go v0.15.7 and later
* Python v12.0.0 and later
* Java v9.0.0 and later
* Javascript v0.10.3 and later
* Ruby master branch
* Haskell v0.3.0.0

We recommend that you update your application and its dependencies to use one of the above client SDK versions if you are on an older version.

While not an exhaustive list, the below AWS components have been updated to use the newer Kubernetes client SDKs that automatically refetches the token :

* Amazon VPC CNI: v1.8.0 and later
* CoreDNS: v1.8.4 and later
* AWS Load Balancer Controller: v2.0.0 and later
* kube-proxy: v1.21.2-eksbuild.2 and later 

You can find instructions for upgrading Amazon VPC CNI, CoreDNS, AWS Load Balancer Controller and kube-proxy in EKS documentation at [3], [4], [5] and [6] respectively. Please check EKS documentation [2] for list of components and the versions that have addressed this issue.

How can I identify service accounts using stale tokens?

As of April 20th 2022, we have identified the below service accounts attached to pods in one or more of your EKS clusters using stale (older than 1 hour) tokens. Service accounts are listed in the format : <eks-cluster-arn>|<namespace:serviceaccount>

arn:aws:eks:XXXXX|istio-system:kiali

You can also use the CloudWatch Log insights query provided in the troubleshooting section of EKS documentation [7] to identify the current list of pods using service accounts with stale tokens and the time elapsed since the token was created.

We recommend that you update your applications and its dependencies that are using stale service accounts tokens to use one of the newer Kubernetes Client SDKs that refetches tokens.

If the service account token used is close to expiry (<90 days) and you do not have sufficient time to update your client SDK versions before expiry, then you can terminate existing pods and create new ones. This results in refetching of the service account token, giving you additional time (90 days) to update your client SDKs.

If you have any questions or concerns, please reach out to AWS Support [8].

[1] https://github.com/kubernetes/enhancements/issues/542
[2] https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html#kubernetes-1.21
[3] https://docs.aws.amazon.com/eks/latest/userguide/managing-vpc-cni.html
[4] https://docs.aws.amazon.com/eks/latest/userguide/managing-coredns.html
[5] https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html
[6] https://docs.aws.amazon.com/eks/latest/userguide/managing-kube-proxy.html
[7] https://docs.aws.amazon.com/eks/latest/userguide/troubleshooting.html#troubleshooting-boundservicetoken
[8] https://aws.amazon.com/support

Sincerely,
Amazon Web Services

**Comment by nrfox** [comment]
Thanks for opening this issue. I suspect this is more related to how Kiali uses its service account token and less to do with the particular version of kubeclient Kiali is using. Kiali reads the service account token from a file and then saves it for further use. It uses this token to talk to kube and can use it to talk to some external services like Prometheus. Perhaps one of those use cases assumes that the token doesn't expire which is a problem if the service account token does expire. @israel-hdez or @lucasponce wdyt?

Specifically using the cached token here: https://github.com/kiali/kiali/blob/master/kubernetes/token.go#L10 could be a problem if that token is never refreshed.

**Comment by ldemailly** [comment]
Thanks! That would explain not finding anything wrong in go.mod and seems like the right analysis indeed, thanks @nrfox

**Comment by israel-hdez** [comment]
This looks related to Istio's preference of using "[Third party service account tokens](https://istio.io/latest/docs/ops/best-practices/security/#configure-third-party-service-account-tokens)", which is a Kubernetes feature that has been available in kubernetes for a while.

Yes, I think it's more about how Kiali is using the ServiceAccount token rather than the client library. To me, this is an indication  that short-lived ServiceAccounts tokens are being preferred over the traditional never-expiring SA tokens.

> Perhaps one of those use cases assumes that the token doesn't expire which is a problem if the service account token does expire.

The only that comes comes into my mind is the cache, which uses watches and holds a copy of the SA token at initialization... After initialization is done, AFAIK, it never re-reads the SA token.

**Comment by bboerst-roku** [comment]
We just saw this issue in our 1.21 cluster where the age of our Kiali pod reached 90d, which caused our service account token to become 'stale' and seeing a flood of Kiali logs like:
```
Failed to watch *v1beta1.RequestAuthentication: failed to list *v1beta1.RequestAuthentication: Unauthorized
Failed to watch *v1alpha3.ServiceEntry: failed to list *v1alpha3.ServiceEntry: Unauthorized
Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Unauthorized
Failed to watch *v1alpha3.VirtualService: failed to list *v1alpha3.VirtualService: Unauthorized
Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Unauthorized
Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
Failed to watch *v1.Deployment: failed to list *v1.Deployment: Unauthorized
Failed to watch *v1.DaemonSet: failed to list *v1.DaemonSet: Unauthorized
Failed to watch *v1.Pod: failed to list *v1.Pod: Unauthorized
Failed to watch *v1alpha3.VirtualService: failed to list *v1alpha3.VirtualService: Unauthorized
Failed to watch *v1alpha3.ServiceEntry: failed to list *v1alpha3.ServiceEntry: Unauthorized
Failed to watch *v1beta1.PeerAuthentication: failed to list *v1beta1.PeerAuthentication: Unauthorized
Failed to watch *v1alpha3.Gateway: failed to list *v1alpha3.Gateway: Unauthorized
Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Unauthorized
Failed to watch *v1alpha3.ServiceEntry: failed to list *v1alpha3.ServiceEntry: Unauthorized
Failed to watch *v1alpha3.WorkloadEntry: failed to list *v1alpha3.WorkloadEntry: Unauthorized
Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Unauthorized
Failed to watch *v1alpha3.Sidecar: failed to list *v1alpha3.Sidecar: Unauthorized
```

Restarting the pod recycled the 90d token and got us back up and running.

This is ref to the Kubernetes change, describing the expiration of service account tokens introduced in 1.21: https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html#kubernetes-1.21

**Comment by bboerst** [comment]
We just saw this issue in our 1.21 cluster where the age of our Kiali pod reached 90d, which caused our service account token to become 'stale' and seeing a flood of Kiali logs like:
```
Failed to watch *v1beta1.RequestAuthentication: failed to list *v1beta1.RequestAuthentication: Unauthorized
Failed to watch *v1alpha3.ServiceEntry: failed to list *v1alpha3.ServiceEntry: Unauthorized
Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Unauthorized
Failed to watch *v1alpha3.VirtualService: failed to list *v1alpha3.VirtualService: Unauthorized
Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Unauthorized
Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
Failed to watch *v1.Deployment: failed to list *v1.Deployment: Unauthorized
Failed to watch *v1.DaemonSet: failed to list *v1.DaemonSet: Unauthorized
Failed to watch *v1.Pod: failed to list *v1.Pod: Unauthorized
Failed to watch *v1alpha3.VirtualService: failed to list *v1alpha3.VirtualService: Unauthorized
Failed to watch *v1alpha3.ServiceEntry: failed to list *v1alpha3.ServiceEntry: Unauthorized
Failed to watch *v1beta1.PeerAuthentication: failed to list *v1beta1.PeerAuthentication: Unauthorized
Failed to watch *v1alpha3.Gateway: failed to list *v1alpha3.Gateway: Unauthorized
Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Unauthorized
Failed to watch *v1alpha3.ServiceEntry: failed to list *v1alpha3.ServiceEntry: Unauthorized
Failed to watch *v1alpha3.WorkloadEntry: failed to list *v1alpha3.WorkloadEntry: Unauthorized
Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Unauthorized
Failed to watch *v1alpha3.Sidecar: failed to list *v1alpha3.Sidecar: Unauthorized
```

Restarting the pod recycled the 90d token and got us back up and running.

This is ref to the Kubernetes change, describing the expiration of service account tokens introduced in 1.21: https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html#kubernetes-1.21

**Comment by jmazzitelli** [comment]
Looks like we are going to need some kind of token refresh in the kiali k8s client code.

**Comment by Xunzhuo** [comment]
If no one is available, I would like to help do this in the next week:)

**Comment by israel-hdez** [comment]
@Xunzhuo Yes! You can take this issue.

**Comment by smauermann** [comment]
Heya, is this issue still being worked on? I haven't seen any updates for some time now. We are on EKS and don't really have a choice here but to use time-bound SA tokens.

**Comment by lucasponce** [comment]
Thanks a lot, @Xunzhuo. 
@smauermann: Kiali is a growing community and you are also welcome if you'd like also to help us with some Pull Requests.

**Comment by georgiiturevich-uw** [comment]
@Xunzhuo is there any progress on this issue?

Btw, a similar issue https://github.com/istio/istio/issues/38077 was recently fixed in Istio here https://github.com/istio/istio/pull/39801

**Comment by smauermann** [comment]
Thank you @Xunzhuo! No pressure, I was just checking the status 😄 

@lucasponce I could have a look, but my application coding skills are a bit rusty.

**Comment by jshaughn** [comment]
@Xunzhuo has been very busy lately, @josunect has agreed to take a look at this issue.

**Comment by josunect** [comment]
Yes! I'm looking at this at the moment

**Comment by nrfox** [comment]
The fix for this went out with v1.56 of Kiali if folks want to upgrade their version of Kiali.

**Comment by israel-hdez** [comment]
@ldemailly We generally don't do back-ports unless a CVE is involved.

Currently, Kiali is targeting Istio v1.15. It's simply because it is untested in older versions. Perhaps, just try and see if it behaves OK with older Istio versions. Istio v1.13 is somewhat old and most likely, there will be issues. Istio v1.14 is probably OK, although we cannot tell with confidence.
